server/src/main/java/org/elasticsearch/index/mapper/MapperService.java

```diff
@@ -229,6 +232,24 @@ public DocumentParser documentParser() {
         }
     }
 
+    /**
+     * Parses the mappings (formatted as JSON) into a map
+     */
+    public static Map<String, Object> parseMapping(NamedXContentRegistry xContentRegistry, CompressedXContent mappingSource)
+        throws IOException {
+        try (
+            InputStream in = CompressorFactory.COMPRESSOR.threadLocalInputStream(mappingSource.compressedReference().streamInput());
```

Comment:
Out of curiosity: does this operation involve decompressing the mapping source now? This probably would mean slightly more time spent here in the future, I guess that is negligable but would like to understand the tradeoffs if there are any.

Reply:
Yea it is in fact decompressing now.
This won't cause any slow-down though I think. In all the potentially hot spots this is used I added it to get around a `CompressedXContent` -> `String` transformation so we're faster there in all cases since we were paying for decompression anyway. Plus this sets up follow-ups that save even more string <-> xcontent conversions.
But even outside of that I don't think it'll even be slower than parsing a string necessarily. We use a thread-local cache for the deflater behind the scenes so setting up the decompression is cheap. When parsing from a string we have all the slowness from parsing `char` instead of `byte`.